core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -790,7 +817,42 @@ class TimeSeriesShard(val ref: DatasetRef,
     shardStats.purgedPartitions.increment(numDeleted)
   }
 
-  def createFlushTask(flushGroup: FlushGroup): Task[Response] = {
+  /**
+    * Creates zero or more flush tasks (up to the number of flush groups) based on examination
+    * of the record container's ingestion time. This should be called before ingesting the container.
+    *
+    * Note that the tasks returned by this method aren't executed yet. The caller decides how
+    * to run the tasks, and by which threads.
+    */
+  def createFlushTasks(container: RecordContainer): Seq[Task[Response]] = {
+    val tasks = new ArrayBuffer[Task[Response]]()
+
+    var oldTimestamp = lastIngestionTime
+    val ingestionTime = Math.max(oldTimestamp, container.timestamp) // monotonic clock
+    var newTimestamp = ingestionTime
+
+    if (newTimestamp > oldTimestamp && oldTimestamp != Long.MinValue) {
+      for (group <- 0 until numGroups optimized) {
+        if (oldTimestamp / flushBoundaryMillis != newTimestamp / flushBoundaryMillis) {
+          // Flush out the group before ingesting records for a new hour (by group offset).
+          tasks += createFlushTask(prepareFlushGroup(group))
+        }
+        oldTimestamp -= flushOffsetMillis
+        newTimestamp -= flushOffsetMillis
```

Comment:
I find it hard to understand the logic quickly. Conversion to something like this or more docs would help.

```
0.to(numGroups).filter(  <someCondition> ).map { group =>
   createFlushTask(prepareFlushGroup(group))
}
```

Reply:
```for { g <- 0 to numGroups optimized 
        if   <some condition>
} yield {
     createFlushTask(prepareFlushGroup(group))
}```

Reply:
The subtraction is the confusing part. Changing the style of the code won't eliminate that aspect. The <someCondition> would need to depend on the stateful changing of the timestamps at each iteration.